test.bash

```diff
@@ -0,0 +1,188 @@
+die() {
+	local rc=${1}
+	shift
+	echo "${*}" 1>&2
+	[[ -n ${AUTOSIGN_TMPDIR} ]] && echo "Testing tmpdir is ${AUTOSIGN_TMPDIR}" 1>&2
+	exit "${rc}"
+}
+
+if [[ -z ${LOCAL_KEYSERVER} && -z ${AUTOSIGN_NO_SEND_KEYS} ]]; then
+	die 1 "Please set LOCAL_KEYSERVER to local keyserver address for pushing, or AUTOSIGN_NO_SEND_KEYS=1 to disable pushing."
+fi
+
+TEST_RC=0
+
+# For testing, keep a single tmpdir, and iterate phases.
+AUTOSIGN_TMPDIR=$(mktemp -d /tmp/autosign-test.XXXXXX) || die 2 "Failed to create testing directory"
+GNUPGHOME=${AUTOSIGN_TMPDIR}/gnupghome
+export AUTOSIGN_TMPDIR GNUPGHOME
+echo "Testing tmpdir is ${AUTOSIGN_TMPDIR}"
+
+umask 077 && mkdir -p "${GNUPGHOME}"
+[[ -n ${AUTOSIGN_NO_SEND_KEYS} ]] || echo "keyserver ${LOCAL_KEYSERVER}" >"${GNUPGHOME}"/gpg.conf
+
+# This is a complete hack.
+DO_REVOKE_TEST=0
+if [[ -z "${AUTOSIGN_FILTER}" ]]; then
+	: "${AUTOSIGN_FILTER:=(&(gentooStatus=active)(gpgfingerprint=*))}"
+	DO_REVOKE_TEST=1
+	# All devs
+	FILTER_00="${AUTOSIGN_FILTER}"
+	# Exclude a few devs to trigger revoke.
+	FILTER_01="(&${AUTOSIGN_FILTER}(!(uid=a*)))"
+	AUTOSIGN_FILTER=${FILTER_00}
+	export AUTOSIGN_FILTER
+fi
+
+authuid_foo='Testing Authority Key Foo <auth-foo@example.com>'
+authuid_bar='Testing Authority Key Bar <auth-bar@example.com>'
+
+echo "Generating dummy keys (set 1)..."
+# this is to ensure there are other secret keys, and we aren't just lucky with the
+# code picking a random key that was the correct one for the key.
+for f in {1..4}; do
+	gpg --batch --passphrase '' --quick-gen-key "Dummy Key 0x$(printf %04x "${f}")"
+done
+
+gpg_colon_get_sec_fpr() {
+	# relies on the ordering that fingerprints of a key are strictly listed after it.
+	awk -F: '($1 == "sec"){sec_k=$5 }($1 == "fpr"){fpr=$10; r= sec_k "$" ; if(match(fpr, r)){ print fpr}}'
+}
+
+echo "Generating authority keys ..."
+gpg -q --batch --passphrase '' --quick-gen-key "${authuid_foo}"
+authfpr_foo=$(
+	gpg --with-colon --list-secret-keys "${authuid_foo}" | gpg_colon_get_sec_fpr
+)
+
+gpg -q --batch --passphrase '' --quick-gen-key "${authuid_bar}"
+authfpr_bar=$(
+	gpg --with-colon --list-secret-keys "${authuid_bar}" | gpg_colon_get_sec_fpr
+)
+if [[ ! ${authfpr_bar} =~ ^[A-F0-9]{40}$ ]]; then
+	echo "Failed to generate authority key 'bar', expected HEX{40}, got fingerprint ${authfpr_bar}" 1>&2
+	TEST_RC=1
+fi
+if [[ ! ${authfpr_foo} =~ ^[A-F0-9]{40}$ ]]; then
+	echo "Failed to generate authority key 'foo', expected HEX{40}, got fingerprint ${authfpr_foo}" 1>&2
+	TEST_RC=1
+fi
+[[ ${TEST_RC} -ne 0 ]] && die "${TEST_RC}" ""
+
+echo "Generating dummy keys (set 2)..."
+# this is to ensure there are other secret keys, and we aren't just lucky with the
+# code picking a random key that was the correct one for the key.
+for f in {5..8}; do
+	gpg --batch --passphrase '' --quick-gen-key "Dummy Key <dummy-0x$(printf %04x "${f}")@example.com>"
+done
+
+if [[ -z "${authfpr_foo}" || -z "${authfpr_bar}" || "${authfpr_bar}" == "${authfpr_foo}" ]]; then
```

Comment:
Isn't `-z "${authfpr_foo}" || -z "${authfpr_bar}"` impossible, given the check above?